### PR TITLE
Merge in 2406; Some Lhotse work; Some previous main cleanup

### PR DESCRIPTION
**Pull request by blisc**

### CODE REVIEW

File: nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py [hunk 1]
```diff
@@ -149,20 +150,20 @@ class MegatronT5SpeechLMModel(MegatronBaseSpeechLM):
...
     for other tasks.
 
     P-tuning initializes an LSTM encoder model that generates virtual prompt
     embeddings for every task. Each task shares the same encoder. After p-tuning
-    is complete, the learned virtual prompts can be saved to the prompt table
+    is compelete, the learned virtual prompts can be saved to the prompt table
     using add_ptuned_prompts_to_prompt_table(). Thus, if a user wants to add a
     new virtual prompt via p-tuning, they do not need to retrain on all previous
-    tasks. This gives p-tuning the same task flexibility as prompt-tuning.
+    tasks. This gives p-tuning the same task flexiblity as prompt-tuning.
```
Comment (XuesongYang) [inline nemo/collections/tts/models/speechllm/megatron_t5_speechllm_model.py:166]: nit: fix typos.

File: nemo/collections/tts/models/speechllm/megatron_base_speechllm_prompt_model.py [hunk 2]
```diff
@@ -333,6 +338,11 @@ def setup_training_data(self, training_data_config=None):
                 num_workers=self.cfg.data.num_workers,
                 pin_memory=True,
             )
+        elif self.cfg.data.get('use_lhotse', None):
+            self._train_ds = build_speechllm_dataset(self, self.cfg.data.lhotse_train_ds, True)
+            consumed_samples = self.compute_consumed_samples(0)
+            self._train_dl = build_speechllm_dataloader(self._train_ds, self.cfg.data.lhotse_train_ds, consumed_samples)
```
Comment (XuesongYang) [inline nemo/collections/tts/models/speechllm/megatron_base_speechllm_prompt_model.py:344]: maywe we should raise `ValueError` that shows only supporting `train_ds`, `train_manifest`, and `use_lhoste`, nothing else.

File: nemo/collections/nlp/modules/common/megatron/attention.py [hunk 3]
```diff
@@ -447,18 +447,13 @@ def forward(
...
                     lora_kv_adapter = self.get_adapter_module(AdapterName.LORA_KV_ADAPTER)
                     if lora_kv_adapter and self.adapter_cfg[AdapterName.LORA_KV_ADAPTER]['enabled']:
                         lora_mixed_kv_layer = lora_kv_adapter(encoder_output)
                         mixed_kv_layer = mixed_kv_layer + lora_mixed_kv_layer
-            mixed_kv_layer, _ = self.key_value(encoder_output)
-            if self.is_adapter_available():
-                lora_kv_adapter = self.get_adapter_module(AdapterName.LORA_KV_ADAPTER)
-                if lora_kv_adapter and self.adapter_cfg[AdapterName.LORA_KV_ADAPTER]['enabled']:
-                    lora_mixed_kv_layer = lora_kv_adapter(encoder_output)
-                    mixed_kv_layer = mixed_kv_layer + lora_mixed_kv_layer
```
Comment (XuesongYang) [inline nemo/collections/nlp/modules/common/megatron/attention.py:461]: This snippet seems odd. In 2402, they are not exisitng. pls double check if 2406 has added something. But at a glance, removing is correct.

File: nemo/collections/multimodal/speech_llm/parts/utils/data_utils.py [hunk 4]
```diff
@@ -380,3 +382,52 @@ def _process_example(self, context: str, output: str):
...
+    def __init__(self, tokenizer, lm_vocab_size, g2p_conf):
+        self.tokenizer = tokenizer
+        self.lm_vocab_size = tokenizer.vocab_size if lm_vocab_size is None else lm_vocab_size
+        self.phoneme_tokenizer = instantiate(g2p_conf)
+        self.pad_id = tokenizer.pad_id
+        self.bos_id = tokenizer.bos_id
+        self.eos_id = tokenizer.eos_id
+
+    def _process_example(self, context: str, output: str):
+        print(context)
```
Comment (XuesongYang) [inline nemo/collections/multimodal/speech_llm/parts/utils/data_utils.py:397]: nit: replace `print` with `logging.debug()`, or just remove it.

File: nemo/collections/multimodal/speech_llm/parts/utils/data_utils.py [hunk 5]
```diff
@@ -380,3 +382,52 @@ def _process_example(self, context: str, output: str):
...
+        check_len = max(len("Text to speech this"), len("Phoneme TTS"))+5
```
Comment (XuesongYang) [inline nemo/collections/multimodal/speech_llm/parts/utils/data_utils.py:398]: the instruction text are still hardcoded inside member functions. Maybe we should declare the instructions when initializing the dataset so that they can be re-used anywhere else. for example in `__init__`:
```python
        self.instructions = ["Phoneme TTS", "Text to speech this"]
```

Comment (XuesongYang) [reply nemo/collections/multimodal/speech_llm/parts/utils/data_utils.py:398]: and again, it worth to add a comment showing why `+5`?

File: nemo/collections/multimodal/speech_llm/parts/utils/data_utils.py [hunk 6]
```diff
@@ -380,3 +382,52 @@ def _process_example(self, context: str, output: str):
...
+        if "Phoneme TTS" in context[:check_len]:
+            instruction = self._get_text_tokens("Phoneme TTS ")
+            context = context.replace("Phoneme TTS", "").strip()
```
Comment (XuesongYang) [inline nemo/collections/multimodal/speech_llm/parts/utils/data_utils.py:401]: this was a known issue that has been fixed already in https://github.com/blisc/NeMo/pull/32. Please cherry-pick the changes.

File: nemo/collections/multimodal/speech_llm/parts/utils/data_utils.py [hunk 7]
```diff
@@ -380,3 +382,52 @@ def _process_example(self, context: str, output: str):
...
+            raise ValueError(f"Did not find TTS this nor PhonemeTTS in {context}")
+        final_question.append(self.eos_id)
+        return torch.tensor(final_question)
+
+    def _get_text_tokens(self, text):
+        input_ids = self.tokenizer.text_to_ids(text)
+        return input_ids
+
+    def _get_phoneme_tokens(self, text, lang="en"):
+        # if self.english_only_model:
```
Comment (XuesongYang) [inline nemo/collections/multimodal/speech_llm/parts/utils/data_utils.py:418]: remove the comment.

File: nemo/collections/multimodal/speech_llm/parts/utils/data_utils.py [hunk 8]
```diff
@@ -380,3 +382,52 @@ def _process_example(self, context: str, output: str):
...
+        input_ids_adjusted = self.tokenizer.text_to_ids("".join(input_ids_adjusted))
+        return input_ids_adjusted
+        # else:
+        #     text = any_locale_text_preprocessing(text)
+        #     input_ids = self.g2p[lang](text)
+        #     input_ids_adjusted = []
+        #     for i in input_ids:
+        #         input_ids_adjusted.append(f"p{{{i}}}")
+        #     input_ids_adjusted = self.tokenizer.text_to_ids("".join(input_ids_adjusted))
+        #     return input_ids_adjusted
```
Comment (XuesongYang) [inline nemo/collections/multimodal/speech_llm/parts/utils/data_utils.py:432]: remove the comments.

File: nemo/collections/common/data/lhotse/dataloader.py [hunk 9]
```diff
@@ -193,7 +193,7 @@ def get_lhotse_dataloader_from_config(
         cuts = cuts.map(partial(_select_channel, channel_selector=config.channel_selector))
 
     # Resample as a safeguard; it's a no-op when SR is already OK
-    cuts = cuts.resample(config.sample_rate)
+    # cuts = cuts.resample(config.sample_rate)
```
Comment (XuesongYang) [inline nemo/collections/common/data/lhotse/dataloader.py:196]: is it on purpose to comment it out?

File: nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py [hunk 10]
```diff
@@ -39,23 +48,10 @@ def collate_vectors(items, max_length: int, padding_value):
...
-    by Lhotse samplers instead.
-
-    Args:
-        text_processor: TextProcessing object
-        default_context: Default question to use if no question is provided
-        tokens_to_generate: Number of tokens to generate during inference
-        pad_to_max_length: Whether to pad the input to the max sequence length. If False, will pad to the max length of the current batch.
-        max_seq_length: Maximum sequence length for each dataset examples. Examples will either be truncated to fit this length or dropped if they cannot be truncated.
-        context_key: Key to use for the context in your JSONL file
-        default_context_key: Key to use for the default context in lhotse yaml
```
Comment (XuesongYang) [inline nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:58]: it would be still great to add some docstring here if not all.

File: nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py [hunk 11]
```diff
@@ -78,49 +78,298 @@ def __init__(
         self.default_context = default_context
         self.context_key = context_key
         self.default_context_key = default_context_key
+        self.task_templates = task_templates
+        self.cross_attention_epsilon = cross_attention_epsilon
+        self.use_attention_prior = True
+        self.context_conditioning ="decoder"
+        self.decoder_context_len = 3*86
```
Comment (XuesongYang) [inline nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:85]: these two numbers 3 and 86 should not be hardcoded.

File: nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py [hunk 12]
```diff
@@ -78,49 +78,298 @@ def __init__(
...
+            split_end = total_inserted_tokens + virtual_token_splits[idx]
+            pseudo_tokens_for_split = "".join(self.pseudo_tokens[split_start:split_end])
+            input_example = input_example.replace(f'<|VIRTUAL_PROMPT_{idx}|>', pseudo_tokens_for_split)
+            total_inserted_tokens = split_end
+
+        return input_example
+
+    def get_position_ids(self, context_and_question):
+        enc_input = []
+        # enc_input.append(virtual_token)
```
Comment (XuesongYang) [inline nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:105]: remove comments

File: nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py [hunk 13]
```diff
@@ -78,49 +78,298 @@ def __init__(
...
+        enc_input = torch.cat(enc_input, dim=1)
+
+        enc_input_p = enc_input[:, 0, :] if enc_input.dim() == 3 else enc_input
+        return build_position_ids(enc_input_p).contiguous()
 
     def __getitem__(self, cuts) -> dict[str, Union[torch.Tensor, list[str], dict]]:
         cuts = cuts.sort_by_duration()
+        cuts_tts = cuts
+        print(f"{len(cuts)}")
+        # cuts_asr = []
```
Comment (XuesongYang) [inline nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:120]: remove comments.

File: nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py [hunk 14]
```diff
@@ -78,49 +78,298 @@ def __init__(
...
+        #     for _, cut in enumerate(cuts):
+        #         if hasattr(cut, self.context_key):
+        #             cut.context = getattr(cut, self.context_key)
+        #         elif hasattr(cut, self.default_context_key):
+        #             cut.context = getattr(cut, self.default_context_key)
+        #         else:
+        #             cut.context = self.default_context
+
+        #     metadata = []
+        #     for id, cut in enumerate(cuts):
```
Comment (XuesongYang) [inline nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:153]: it seems there are major changes in this file. Could you remove all comments?

File: examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+  max_position_embeddings: ${model.data.max_seq_length}
+
+  frozen_model:
+    # micro_batch_size: null
+    # global_batch_size: null
+    # megatron_amp_O2: true
+    # seq_length: 512
+    # max_position_embeddings: 512
+    # precision: bf16
+    # Above is overridden in code
```
Comment (XuesongYang) [inline examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml:83]: remove comments.

File: examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+    seed: 1234
+    use_cpu_initialization: false
+    apex_transformer_log_level: 30
+    tokenizer:
+      library: megatron
+      type: BertWordPieceCase
+      model: null
+      vocab_file: null
+      merge_file: null
+      # num_sentinel_tokens: 100
```
Comment (XuesongYang) [inline examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml:103]: remove comments.

File: examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+    fp16_lm_cross_entropy: false
+    seed: 1234
+    use_cpu_initialization: false
+    apex_transformer_log_level: 30
+    tokenizer:
+      library: megatron
+      type: BertWordPieceCase
+      model: null
+      vocab_file: null
+      merge_file: null
```
Comment (XuesongYang) [inline examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml:102]: i remembered in our NeMo policy, we should use `???` in yaml config rather than `null` i f it requires users to explicitly specify the file path.

File: examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+      # num_sentinel_tokens: 100
+    optim:
+      name: null
```
Comment (XuesongYang) [inline examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml:105]: same here, `null` --> `???`

File: examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml [hunk 19]
```diff
@@ -0,0 +1,307 @@
...
+      # {'audio_filepath': 'audio1.wav', 'offset': 0.0, 'duration': 12.3, 'question': 'transcribe this audio', 'answer': 'I have a dream...'}
+      # the 'answer' field can also be 'text', and a default 'question' field is added if missing in manigests, so as to work with ASR manifests
+      input_cfg:
+      # - type: nemo_tarred
+      #   manifest_filepath: /home/jasoli/data_prime/ASR/Suno/yt10k_en/sharded_manifests/bucket1/manifest__OP_0..511_CL__punctuation_removed.json
+      #   tarred_audio_filepaths: /home/jasoli/data_prime/ASR/Suno/yt10k_en/bucket1/audio__OP_0..511_CL_.tar
+      #   weight: 0.5
+      #   tags:
+      #     lang: en
+      #     pnc: no
```
Comment (XuesongYang) [inline examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml:171]: remove comments

File: examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml [hunk 20]
```diff
@@ -0,0 +1,307 @@
...
+      g2p_conf:
+        _target_: nemo.collections.tts.g2p.models.i18n_ipa.IpaG2p
+        phoneme_dict: "scripts/tts_dataset_files/ipa_cmudict-0.7b_nv23.01.txt"
+        heteronyms: "scripts/tts_dataset_files/heteronyms-052722"
+        phoneme_probability: 0.8
+        ignore_ambiguous_words: False
+        use_chars: True
+        use_stresses: True
+        grapheme_prefix: ${model.data.grapheme_prefix}
+      # sample_alpha: 0.1
```
Comment (XuesongYang) [inline examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml:232]: remove comments

File: examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml [hunk 21]
```diff
@@ -0,0 +1,307 @@
...
+    sample_rate: 24000
+    add_eos: true
+    add_bos: false
+    use_attention_prior: true
+    attention_prior_scaling_factor: 0.05
+    cross_attention_epsilon: 0.0
+    decoder_starts_with_pad: False
+    add_eos_to_decoder_output: True
+    add_sentinel_to_input: True
+    ul2_prompt_token: null # <extra_id_s>, <extra_id_r>, <extra_id_x>
```
Comment (XuesongYang) [inline examples/tts/speechllm/conf/megatron_t5_speechlm_model_lhotse.yaml:244]: should we put `null` here? or perhaps, better to remove it and add `None` as default value in the code?

**Review by XuesongYang — CHANGES_REQUESTED** [review]
Had a glance of code overall. Please remove all comments that do not describe the codes. It would be also great if could add some unit tests showcasing how the lhotse dataloader works and avoiding corner cases.
